benchmarks/distributed/tensor/tp_benchmark.py

```diff
@@ -0,0 +1,247 @@
+    ) -> "DummyDTensor":
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls,
+            local_tensor.size(),
+            strides=local_tensor.stride(),
+            dtype=local_tensor.dtype,
+            device=local_tensor.device,
+            layout=local_tensor.layout,
+            requires_grad=local_tensor.requires_grad,
+        )
+        r._local_tensor = local_tensor
+        return r
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap_schema(e: object) -> object:
+            return e._local_tensor if isinstance(e, DummyDTensor) else e
+
+        def wrap_schema(e: object) -> object:
+            return DummyDTensor(e) if isinstance(e, torch.Tensor) else e
+        args_unwrap = tree_map(unwrap_schema, args)
+        kwargs_unwrap = tree_map(unwrap_schema, kwargs)
+        return tree_map(wrap_schema, func(*args_unwrap, **kwargs_unwrap))
+
+
+
+class MLPModel(nn.Module):
+    def __init__(self, dim_size):
+        super(MLPModel, self).__init__()
+        self.net1 = nn.Linear(128, dim_size)
+        self.relu = nn.ReLU()
+        self.net2 = nn.Linear(dim_size, 128)
+
+    def forward(self, x):
+        return self.net2(self.relu(self.net1(x)))
+
+
+
+
+@contextlib.contextmanager
+def maybe_run_profiler(args, path):
+    if args.dump_profiler:
+        with torch.profiler.profile(
+            activities=[
+                torch.profiler.ProfilerActivity.CPU,
+                torch.profiler.ProfilerActivity.CUDA,
+            ],
+            schedule=torch.profiler.schedule(wait=1, warmup=2, active=2, repeat=1),
+            on_trace_ready=torch.profiler.tensorboard_trace_handler(path),
+            profile_memory=True,
+            with_stack=True,
+            record_shapes=True,
+        ) as torch_profiler:
+            yield torch_profiler
+    else:
+        torch_profiler = contextlib.nullcontext()
+        yield torch_profiler
+
+
+def run_tp(rank, args):
+    """
+    Main body of the demo of a basic version of tensor parallel by using
+    PyTorch native APIs.
+    """
+    print(f"Running basic Megatron style TP example on rank {rank}.")
+    setup(rank, args.world_size)
+    tp_degree = args.world_size
+    dim_size = 64
+    # create a sharding plan based on the given world_size.
+    device_mesh = DeviceMesh(
+        "cuda",
+        torch.arange(args.world_size),
+    )
+    LR = 0.25
+
+    # Control group to mimic Megatron.
+    model = MLPModel(dim_size // tp_degree).cuda(rank)
+    optimizer = torch.optim.SGD(model.parameters(), lr=LR)
+    inp = torch.rand(128, 128).cuda(rank)
+    torch.distributed.barrier()
+    with maybe_run_profiler(args, "./control/") as torch_profiler:
+        for i in range(args.iter_nums):
+            if args.sequence_parallel:
+                output = torch.empty(inp.size(0) * tp_degree, *inp.size()[1:], device=inp.device)
+                input = _all_gather_base(output, inp)
```

Comment:
do you need `else: input = inp` to make this run with `--sequence-parallel False`